lib/msal-browser/src/controllers/NestedAppAuthController.ts

```diff
@@ -0,0 +1,525 @@
+import { BrowserConfiguration } from "../config/Configuration";
+import { INavigationClient } from "../navigation/INavigationClient";
+import { AuthorizationCodeRequest } from "../request/AuthorizationCodeRequest";
+import { EndSessionPopupRequest } from "../request/EndSessionPopupRequest";
+import { EndSessionRequest } from "../request/EndSessionRequest";
+import { PopupRequest } from "../request/PopupRequest";
+import { RedirectRequest } from "../request/RedirectRequest";
+import { SilentRequest } from "../request/SilentRequest";
+import { SsoSilentRequest } from "../request/SsoSilentRequest";
+import { ApiId, WrapperSKU, InteractionType } from "../utils/BrowserConstants";
+import { IController } from "./IController";
+import { TeamsAppOperatingContext } from "../operatingcontext/TeamsAppOperatingContext";
+import { IBridgeProxy } from "../naa/IBridgeProxy";
+import { CryptoOps } from "../crypto/CryptoOps";
+import { NestedAppAuthAdapter } from "../naa/mapping/NestedAppAuthAdapter";
+import { NestedAppAuthError } from "../error/NestedAppAuthError";
+import { EventHandler } from "../event/EventHandler";
+import { EventType } from "../event/EventType";
+import { EventCallbackFunction, EventError } from "../event/EventMessage";
+import { AuthenticationResult } from "../response/AuthenticationResult";
+import { BrowserCacheManager } from "../cache/BrowserCacheManager";
+import { ClearCacheRequest } from "../request/ClearCacheRequest";
+
+export class NestedAppAuthController implements IController {
+    // OperatingContext
+    protected readonly operatingContext: TeamsAppOperatingContext;
+
+    // BridgeProxy
+    protected readonly bridgeProxy: IBridgeProxy;
+
+    // Crypto interface implementation
+    protected readonly browserCrypto: ICrypto;
+
+    // Input configuration by developer/user
+    protected readonly config: BrowserConfiguration;
+
+    // Logger
+    protected logger: Logger;
+
+    // Performance telemetry client
+    protected readonly performanceClient: IPerformanceClient;
+
+    // EventHandler
+    protected readonly eventHandler: EventHandler;
+
+    // NestedAppAuthAdapter
+    protected readonly nestedAppAuthAdapter: NestedAppAuthAdapter;
+
+    constructor(operatingContext: TeamsAppOperatingContext) {
+        this.operatingContext = operatingContext;
+        const proxy = this.operatingContext.getBridgeProxy();
+        if (proxy !== undefined) {
+            this.bridgeProxy = proxy;
+        } else {
+            throw new Error("unexpected: bridgeProxy is undefined");
+        }
+
+        // Set the configuration.
+        this.config = operatingContext.getConfig();
+
+        // Initialize logger
+        this.logger = this.operatingContext.getLogger();
+        // Initialize performance client
+        this.performanceClient = this.config.telemetry.client;
+
+        // Initialize the crypto class.
+        this.browserCrypto = operatingContext.isBrowserEnvironment()
```

Comment:
Do you need to sign tokens inside NAA? If not you can remove this and use the encoding/decoding/guid generation functions directly

Reply:
Seems like an optimization that TAOS can pick up.